saleor/payment/gateways/adyen/webhooks.py

```diff
@@ -0,0 +1,654 @@
+        if user_is_correct and check_password(request_password, password):
+            return True
+    except binascii.Error:
+        pass
+    return False
+
+
+def handle_webhook(request: WSGIRequest, gateway_config: "GatewayConfig"):
+    json_data = json.loads(request.body)
+    # JSON and HTTP POST notifications always contain a single NotificationRequestItem
+    # object.
+    notification = json_data.get("notificationItems")[0].get(
+        "NotificationRequestItem", {}
+    )
+
+    if not validate_hmac_signature(notification, gateway_config):
+        return HttpResponseBadRequest("Invalid or missing hmac signature.")
+    if not validate_auth_user(request.headers, gateway_config):
+        return HttpResponseBadRequest("Invalid or missing basic auth.")
+
+    event_handler = EVENT_MAP.get(notification.get("eventCode", ""))
+    if event_handler:
+        event_handler(notification, gateway_config)
+        return HttpResponse("[accepted]")
+    return HttpResponseNotFound()
+
+
+def handle_additional_actions(
+    request: WSGIRequest,
+    payment_details: Callable,
+    adyen_auto_capture: bool,
+    auto_capture: bool,
+):
+    payment_id = request.GET.get("payment")
+    checkout_pk = request.GET.get("checkout")
+
+    if not payment_id or not checkout_pk:
+        return HttpResponseNotFound()
+
+    _type, payment_pk = from_global_id(payment_id)
+    try:
+        payment = Payment.objects.get(
+            pk=payment_pk,
+            checkout__pk=checkout_pk,
+            is_active=True,
+            gateway="mirumee.payments.adyen",
+        )
+    except ObjectDoesNotExist:
+        return HttpResponseNotFound(
+            "Cannot perform payment. "
+            "There is no active adyen payment with specified checkout."
+        )
+
+    extra_data = json.loads(payment.extra_data)
+    data = extra_data[-1] if isinstance(extra_data, list) else extra_data
+
+    return_url = payment.return_url
+
+    if not return_url:
+        return HttpResponseNotFound(
+            "Cannot perform payment. Lack of data about returnUrl."
+        )
+
+    try:
+        request_data = prepare_api_request_data(request, data)
+    except KeyError as e:
+        return HttpResponseBadRequest(e.args[0])
+    try:
+        result = api_call(request_data, payment_details)
+    except PaymentError as e:
+        return HttpResponseBadRequest(str(e))
+
+    handle_api_response(payment, result, adyen_auto_capture, auto_capture)
+
+    redirect_url = prepare_redirect_url(payment_id, checkout_pk, result, return_url)
+    return redirect(redirect_url)
+
+
+def prepare_api_request_data(request: WSGIRequest, data: dict):
+    if "parameters" not in data or "payment_data" not in data:
+        raise KeyError(
+            "Cannot perform payment. Lack of payment data and parameters information."
+        )
+
+    params = data["parameters"]
+    request_data: "QueryDict" = QueryDict("")
+
+    if all([param in request.GET for param in params]):
```

Comment:
Unless we really need to check presence of all param keys in request, a more idiomatic way to check if request is POST or not is:
```
if request.method == "POST":
    request_data = request.POST
else:
    request_data = request.GET
```

Reply:
We need to confirm that we get all required parameters to be able to proceed to the next step.